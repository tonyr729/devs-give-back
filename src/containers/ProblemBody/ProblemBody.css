.welcome {
  font-size: 2rem;
  padding-top: 10rem; }

.body-instructions {
  font-size: 4rem;
  padding-top: 4rem;
  color: #f8f8ff; }

.body-input {
  background-color: rgba(0, 0, 0, 0);
  color: #ffaaaa;
  border: none;
  margin: 4rem auto 0;
  display: block;
  width: 54rem;
  font-size: 2rem;
  height: 20rem;
  text-align: center; }
  .body-input::placeholder {
    color: #ffaaaa91; }
  .body-input:focus {
    outline: none; }

.button-container {
  display: flex;
  width: 10rem;
  height: 5rem;
  margin: 0 auto;
  background-color: #607d8b; }

.next-button,
.previous-button {
  width: 6rem;
  height: 6rem;
  background: none;
  border: none;
  display: block;
  margin: 0rem auto; }
  .next-button img,
  .previous-button img {
    width: 4rem; }
