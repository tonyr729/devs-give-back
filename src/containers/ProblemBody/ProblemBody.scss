$grey: #4f4f4f;
$white: #ffffff;
$green: #39bda7;
$blue: #03a9f4;
$transparent: rgba(255, 255, 255, 0);
$black-05: rgba(0, 0, 0, 0.05);

.welcome {
  font-size: 2rem;
  padding-top: 10rem;
}

.body-instructions {
  font-size: 4rem;
  padding-top: 4rem;
  color: $grey;
}


.body-input {
  background-color: $transparent;
  color: $blue;
  border: 1px solid $blue;
  margin: 4rem auto 0;
  display: block;
  width: 54rem;
  font-size: 2rem;
  height: 20rem;
  resize: none;
  &::placeholder {
    color: rgba(3, 168, 244, 0.342);;
  }
  &:focus {
    outline: none;
    border: none;
  }
}

.button-container {
  display: flex;
  width: 18rem;
  height: 5rem;
  margin: 0 auto;
  background-color: $white;
}

.next-button,
.previous-button {
  background:$blue;
  color:$white;
  border:none;
  margin: 0 auto;
  position: relative;
  border-radius: 7rem;
  display: block;
  height:60px;
  font-size:1.6em;
  padding:0 2em;
  cursor: pointer;
  transition:800ms ease all;
  outline:none;
  &:hover{
    background:$transparent;
    color:#002e42;
  }
  &:before,
  &:after {
    content:'';
    position:absolute;
    top:0;
    right:0;
    height:2px;
    width:0;
    background: $blue;
    transition:400ms ease all;
  }
  &:after {
    right:inherit;
    top:inherit;
    left:0;
    bottom:0;
  }
  &:hover:before,
  &:hover:after {
    width:100%;
    transition:800ms ease all;
  }  
}

