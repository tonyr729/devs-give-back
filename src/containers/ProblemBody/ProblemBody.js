import React, { Component } from 'react';
import './ProblemBody.css';
import { connect } from 'react-redux';
import { createProblemBody } from '../../actions/actions';
import rightArrow from '../../images/right_arrow.svg';
import leftArrow from '../../images/left_arrow.svg';


class ProblemBody extends Component {
  constructor() {
    super();
    this.state = {
      input: ''
    }
  }

  handleInputChange= (event) => {
    this.setState({
      input: event.target.value
    })
  }

  handleSubmit = (event) => {
    event.preventDefault();
    this.props.createProblemBody(this.state.input);
    this.props.history.push("/problem-category")
  }

  addInput = (event) => {
    if (event.keyCode === 13 && event.shiftKey === false) {
      this.handleSubmit(event);
    }
  }

  logInCheck = (client) => {
    const value = Object.keys(client).length;
    if ( value === 0 ) {
      this.props.history.push("/client-login");
    }
  }
  

  render() {
    this.logInCheck(this.props.client)
    const name = this.props.client.name || 'client unknown';
    const firstName = name.split(' ')[0];

    return (
      <div className='client-background'>
        <p className='welcome'>Sounds great {firstName},</p>
        <p className='body-instructions'>Now give us some more detail</p>
        <textarea className='body-input' onKeyDown={this.addInput} onChange={this.handleInputChange} maxLength='400' value={this.state.input} autoFocus/>
        <div className="button-container">
          <button className='previous-button' onClick={()=> this.props.history.push("/problem-title")} ><img className='left-arrow-img' src={leftArrow} /></button>
          <button className='next-button' onClick={this.handleSubmit} ><img className='right-arrow-img' src={rightArrow} /></button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state) => ({
  client: state.client,
  clientID: state.problemClient,
  title: state.problemTitle,
  body: state.problemBody
});

const mapDispatchToProps = (dispatch) => ({
  createProblemBody: (body) => dispatch(createProblemBody(body))
});


export default connect(mapStateToProps, mapDispatchToProps)(ProblemBody);