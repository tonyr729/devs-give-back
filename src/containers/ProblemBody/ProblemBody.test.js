import { ProblemBody, mapStateToProps, mapDispatchToProps} from './ProblemBody';
import React from 'react';
import { shallow } from 'enzyme';


describe('ProblemBody', () => {
  it('should match snapshot', () => {
    const mockProps = {
      client: {name: 'tony'}
    };
    const wrapper = shallow(<ProblemBody {...mockProps} />);

    expect(wrapper).toMatchSnapshot();
  });

  describe('handleInputChange', () => {
    let mockProps;
    let mockEvent;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'}
      };
      mockEvent = {
        target: {
          value: 'input value'
        }
      }
      wrapper = shallow(<ProblemBody {...mockProps} />);
    });

    it('should setState off the params', () => {
      wrapper.instance().handleInputChange(mockEvent)
    
      expect(wrapper.state('input')).toEqual(mockEvent.target.value)
    });
  });

  describe('handleSubmit', () => {
    let mockProps;
    let mockEvent;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        createProblemBody: jest.fn(),
        history: {
          push: jest.fn()
        }
      };
      mockEvent = {
        preventDefault: jest.fn()
      };
      wrapper = shallow(<ProblemBody {...mockProps} />);
    });

    it('should call the given methods', () => {
      wrapper.instance().handleSubmit(mockEvent)
    
      expect(wrapper.instance().props.createProblemBody).toHaveBeenCalled();
      expect(wrapper.instance().props.history.push).toHaveBeenCalled();
    });
  });

  describe('submitInput', () => {
    let mockProps;
    let mockEvent;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        createProblemBody: jest.fn(),
        history: {
          push: jest.fn()
        }
      };
      mockEvent = {
        keyCode: 13,
        shiftKey: false,
        preventDefault: jest.fn()
      };
      wrapper = shallow(<ProblemBody {...mockProps} />);
    });

    it('should call the given methods if keyCodes match', () => {
      wrapper.instance().handleSubmit = jest.fn();
      
      wrapper.instance().submitInput(mockEvent);
    
      expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
    });

    it('should do nothing if keyCodes dont match', () => {
      mockEvent = {
        keyCode: 11,
        shiftKey: true,
        preventDefault: jest.fn()
      };
      wrapper.instance().handleSubmit = jest.fn();

      wrapper.instance().submitInput(mockEvent);
    
      expect(wrapper.instance().handleSubmit).not.toHaveBeenCalled();
    });
  });

  describe('logInCheck', () => {
    let mockProps;
    let mockEvent;
    let wrapper;

    beforeEach(() => {
      mockProps = {
        client: {name: 'tony'},
        createProblemBody: jest.fn(),
        history: {
          push: jest.fn()
        }
      };
      wrapper = shallow(<ProblemBody {...mockProps} />);
    });

    it('should redirect to login if no client', () => {
      wrapper.instance().logInCheck({})
      
      expect(wrapper.instance().props.history.push).toHaveBeenCalledWith("/client-login");
    });
  });

  describe('mapStateToProps', () => {
    it('returns an object with client info', () => {

      const mockState = {
        client: 'value',
        problemClient: 'value',
        problemTitle: 'value',
        problemBody: 'value'
      }

      const expected = {
        body: "value", 
        client: "value", 
        clientID: "value", 
        title: "value"
      }
      const mappedProps = mapStateToProps(mockState);
      expect(mappedProps).toEqual(expected);
      
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch on createCompletedProblem with the correct params', () => {

      const mockDispatch = jest.fn();
      const body = {prob: 'prob'}
      const mappedProps = mapDispatchToProps(mockDispatch);
      const mockAction = {
        type: 'CREATE_PROBLEM_BODY',
        body
      };
      
      mappedProps.createProblemBody(body);
      expect(mockDispatch).toHaveBeenCalledWith(mockAction);
    });
  });
});