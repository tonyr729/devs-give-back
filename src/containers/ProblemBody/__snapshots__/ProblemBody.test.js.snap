// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProblemBody should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ProblemBody
    client={
        Object {
            "name": "tony",
          }
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <p
          className="welcome"
>
          Sounds great 
          tony
          ,
</p>,
        <p
          className="body-instructions"
>
          Now give us some more detail
</p>,
        <textarea
          autoFocus={true}
          className="body-input"
          maxLength="400"
          onChange={[Function]}
          onKeyDown={[Function]}
          value=""
/>,
        <div
          className="button-container"
>
          <button
                    className="previous-button"
                    onClick={[Function]}
          >
                    <img
                              className="left-arrow-img"
                              src="left_arrow.svg"
                    />
          </button>
          <button
                    className="next-button"
                    onClick={[Function]}
          >
                    <img
                              className="right-arrow-img"
                              src="right_arrow.svg"
                    />
          </button>
</div>,
      ],
      "className": "client-background",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Sounds great ",
            "tony",
            ",",
          ],
          "className": "welcome",
        },
        "ref": null,
        "rendered": Array [
          "Sounds great ",
          "tony",
          ",",
        ],
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Now give us some more detail",
          "className": "body-instructions",
        },
        "ref": null,
        "rendered": "Now give us some more detail",
        "type": "p",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autoFocus": true,
          "className": "body-input",
          "maxLength": "400",
          "onChange": [Function],
          "onKeyDown": [Function],
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "textarea",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <button
              className="previous-button"
              onClick={[Function]}
>
              <img
                            className="left-arrow-img"
                            src="left_arrow.svg"
              />
</button>,
            <button
              className="next-button"
              onClick={[Function]}
>
              <img
                            className="right-arrow-img"
                            src="right_arrow.svg"
              />
</button>,
          ],
          "className": "button-container",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                className="left-arrow-img"
                src="left_arrow.svg"
/>,
              "className": "previous-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "left-arrow-img",
                "src": "left_arrow.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "button",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <img
                className="right-arrow-img"
                src="right_arrow.svg"
/>,
              "className": "next-button",
              "onClick": [Function],
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "right-arrow-img",
                "src": "right_arrow.svg",
              },
              "ref": null,
              "rendered": null,
              "type": "img",
            },
            "type": "button",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <p
            className="welcome"
>
            Sounds great 
            tony
            ,
</p>,
          <p
            className="body-instructions"
>
            Now give us some more detail
</p>,
          <textarea
            autoFocus={true}
            className="body-input"
            maxLength="400"
            onChange={[Function]}
            onKeyDown={[Function]}
            value=""
/>,
          <div
            className="button-container"
>
            <button
                        className="previous-button"
                        onClick={[Function]}
            >
                        <img
                                    className="left-arrow-img"
                                    src="left_arrow.svg"
                        />
            </button>
            <button
                        className="next-button"
                        onClick={[Function]}
            >
                        <img
                                    className="right-arrow-img"
                                    src="right_arrow.svg"
                        />
            </button>
</div>,
        ],
        "className": "client-background",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Sounds great ",
              "tony",
              ",",
            ],
            "className": "welcome",
          },
          "ref": null,
          "rendered": Array [
            "Sounds great ",
            "tony",
            ",",
          ],
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Now give us some more detail",
            "className": "body-instructions",
          },
          "ref": null,
          "rendered": "Now give us some more detail",
          "type": "p",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoFocus": true,
            "className": "body-input",
            "maxLength": "400",
            "onChange": [Function],
            "onKeyDown": [Function],
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "textarea",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <button
                className="previous-button"
                onClick={[Function]}
>
                <img
                                className="left-arrow-img"
                                src="left_arrow.svg"
                />
</button>,
              <button
                className="next-button"
                onClick={[Function]}
>
                <img
                                className="right-arrow-img"
                                src="right_arrow.svg"
                />
</button>,
            ],
            "className": "button-container",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  className="left-arrow-img"
                  src="left_arrow.svg"
/>,
                "className": "previous-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "left-arrow-img",
                  "src": "left_arrow.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "button",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": <img
                  className="right-arrow-img"
                  src="right_arrow.svg"
/>,
                "className": "next-button",
                "onClick": [Function],
              },
              "ref": null,
              "rendered": Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "right-arrow-img",
                  "src": "right_arrow.svg",
                },
                "ref": null,
                "rendered": null,
                "type": "img",
              },
              "type": "button",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
